### DISCUSSION TITLE
destroy directive looses access to member variables

### DISCUSSION THREAD

**Discussion by hirasso** [alpinejs/alpine#2692] [7. Bugs (archived)]
Hi there! Not really sure if this is a bug, but it feels like one to me ;)

I am trying to use [the undocumented `destroy()` directive](https://github.com/alpinejs/alpine/discussions/2218) to unbind an event listener that I have registered inside the `init()` directive of my compontent:

Example HTML:

```html
<div class="my-component" x-data="myComponent()">...</div>
<script src="main.js"></script>
```
Example Component:

```js
export const myComponent = () => ({
  init() {
    // saving a reference to my event handler, so that I can remove it later
    this.popStateHandler = this.popStateHandler.bind(this)
    window.addEventListener('popstate', this.popStateHandler)
  }
  popStateHandler(e) {
    // ...
  }
  destroy() {
    console.log(this.popStateHandler) // logs "undefined", so I can't really unregister the handler anymore
  }
})
```

My `main.js`:

```js
import Alpine from 'alpinejs';
import { myComponent } from '../alpine-components/myComponent.js';

window.Alpine = Alpine
Alpine.data('myComponent', myComponent)
Alpine.start()

// I'm removing all instances of my component after a delay, so that I can test the `destroy` functionality:
setTimeout(() => {
  document.querySelectorAll('.my-component').forEach(el => el.remove())
}, 1000);
```

[Our investigation in the discussion](https://github.com/alpinejs/alpine/discussions/2218#discussioncomment-2228980) showed, that `this` seems not to be pointing to the alpine component anymore, when `destroy()` is being called. It might get called too late? Would it be possible to introduce some directive *just before* destroying the component, so that I could do my cleaning up?

**Comment by adamlogic** [comment]
Is this a duplicate of https://github.com/alpinejs/alpine/discussions/2103, or is there a distinction I'm not seeing?

**Reply by hirasso** [reply]
Ah right, that's correct! Should I add the comment [to your's](https://github.com/alpinejs/alpine/discussions/2103#discussioncomment-2228653), that the custom `cleanup` directive also doesn't preserve the scope? ;)
